README.md
# AES-GCM-SIV

[![CircleCI](https://circleci.com/gh/codahale/aes-gcm-siv.svg?style=svg)](https://circleci.com/gh/codahale/aes-gcm-siv)

A Java implementation of [AES-GCM-SIV](https://eprint.iacr.org/2017/168) 
([draft-irtf-cfrg-gcmsiv-08](https://tools.ietf.org/html/draft-irtf-cfrg-gcmsiv-08)), a
nonce-misuse resistant Authenticated Encryption And Data (AEAD) algorithm.

## Is it ready

No, AES-GCM-SIV is still in draft form and hasn't yet been standardized. This library implements the
algorithm described in
[draft-irtf-cfrg-gcmsiv-08](https://tools.ietf.org/html/draft-irtf-cfrg-gcmsiv-08).

## Is it fast

It's very fast. AES-GCM-SIV's performance is largely dependent on hardware support for AES and GCM.
Java 8 added AES-NI support, but only for AES-CBC, and Java 9 will improve GCM performance via
`pclmulqdq` intrinsics, but only for AES-GCM. All things being equal, AES-GCM is slightly faster for
encryption and slightly slower for decryption. 

Here's some benchmark results from a `c4.xlarge` EC2 instance using Java `1.8.0_131-b11`, comparing
`AES/GCM/NoPadding` to AES-GCM-SIV:

```
Benchmark                       Mode  Cnt   Score   Error  Units
Benchmarks.aes_GCM_Decrypt      avgt  200  24.336 ± 0.026  us/op
Benchmarks.aes_GCM_Encrypt      avgt  200  23.570 ± 0.008  us/op
Benchmarks.aes_GCM_SIV_Decrypt  avgt  200  23.154 ± 0.012  us/op
Benchmarks.aes_GCM_SIV_Encrypt  avgt  200  23.106 ± 0.015  us/op
```

Java 9's GCM intrinsics make it roughly 3x faster (`9+181`):

``` 
Benchmark                       Mode  Cnt   Score   Error  Units
Benchmarks.aes_GCM_Decrypt      avgt  200   8.936 ± 0.313  us/op
Benchmarks.aes_GCM_Encrypt      avgt  200   7.490 ± 0.237  us/op
Benchmarks.aes_GCM_SIV_Decrypt  avgt  200  23.439 ± 1.991  us/op
Benchmarks.aes_GCM_SIV_Encrypt  avgt  200  24.462 ± 0.196  us/op
```

Java 11's further optimizations make it roughly 4x faster (`11+28`):

```
Benchmark                       Mode  Cnt   Score   Error  Units
Benchmarks.aes_GCM_Decrypt      avgt    5   5.987 ± 0.272  us/op
Benchmarks.aes_GCM_Encrypt      avgt    5   5.876 ± 0.726  us/op
Benchmarks.aes_GCM_SIV_Decrypt  avgt    5  20.871 ± 3.363  us/op
Benchmarks.aes_GCM_SIV_Encrypt  avgt    5  19.970 ± 1.719  us/op
```

## Why's it good

AES-GCM-SIV is a nonce-misuse resistant AEAD, which means it doesn't fail catastrophically if a
nonce gets re-used. This is a concern for large systems which involve operations with a single key:

> Since a central allocation system for nonces is not operationally viable, random selection of
nonces is the only possibility. AES-GCM’s limit of 2^32 random nonces (per key) suggests that, even
if the system rotated these secret keys daily, it could not issue more than about 50K tokens per
second. However, in order to process DDoS attacks the system may need to sustain issuance of several
hundred million per second.

Unlike AES-GCM or ChaChaPoly1305, AES-GCM-SIV can tolerate _some_ duplicate nonces, but it still has
limits:

> When discussing this work, we found a widespread misunderstanding of the term “nonce-misuse
resistant”. Many people appear to expect the security of a nonce-misuse resistant scheme to be
completely unaffected by the number of times that a nonce is reused. Thus, while it is a convenient
shorthand to distinguish schemes that tolerate repeated nonces (e.g., AES-GCM-SIV) from those that
do not (e.g., AES-GCM and ChaCha20-Poly1305), nonce-misuse resistance is not necessarily a “binary”
property. In particular, it is not binary for the case of AES-GCM-SIV, and as we have shown the
security bounds of AES-GCM-SIV change as the number of repeated nonces varies.
>
> As such, it is important to understand what security is actually guaranteed by nonce-misuse
resistance. First and foremost, nonce-misuse resistant schemes reveal when the same plaintext is
encrypted using the same nonce, and this is well understood. Due to this, some have concluded that
if an application guarantees unique plaintexts, then the same nonce can be safely reused in every
encryption. Although this is true in some sense, it is also true that the security bounds can be
degraded, as we have shown in this paper. Thus, it is not recommended to purposefully use
AES-GCM-SIV with the same nonce (unless the number of encryptions is small enough so that the
quadratic bound is small). We stress that if the same nonce is used always, then AES-GCM-SIV is no
worse than previous schemes; however, AES-GCM-SIV can achieve far better bounds and it is worth
taking advantage of this. 
> 
> We believe that AES-GCM-SIV is well suited to applications where independent servers need to work
with the same key, and where nonce repetition is a real threat. In such cases, it is not possible to
enjoy the better bounds available for encryption schemes that utilize state to ensure unique nonces
in every encryption.

(from ["AES-GCM-SIV: Specification and Analysis"](https://eprint.iacr.org/2017/168))

That said, you should still use randomly generated nonces:

> We stress that nonce-misuse resistant schemes guarantee that if a nonce repeats then the only
security loss is that identical plaintexts will produce identical ciphertexts.  Since this can also
be a concern (as the fact that the same plaintext has been encrypted twice is revealed), we do not
recommend using a fixed nonce as a policy.  In addition, as we show below, better-than-birthday
bounds are achieved by AES-GCM-SIV when nonces repetition is low.  Finally, as shown in [[BHT18]],
there is a great security benefit in the multi- user/multi-key setting when each particular nonce is
re-used by a small number of users only.  We stress that nonce-misuse resistance was never intended
to be used with intentional nonce-reuse; rather, such schemes provide the best possible security in
the event of nonce reuse.  Due to all of the above, it is RECOMMENDED that nonces be randomly
generated.
> 
> Some example usage bounds for AES-GCM-SIV are given below (for these
bounds, the adversary's advantage is always below 2^-32).  For up to
256 repeats of a nonce (i.e., where one can assume that nonce misuse
is no more than this bound), the following message limits should be
respected (this assumes a short AAD):
>   
> * 2^29 messages, where each plaintext is at most 1GiB
> * 2^35 messages, where each plaintext is at most 128MiB
> * 2^49 messages, where each plaintext is at most 1MiB
> * 2^61 messages, where each plaintext is at most 16KiB

[BHT18]: https://tools.ietf.org/html/draft-irtf-cfrg-gcmsiv-08#ref-BHT18

(from [draft-irtf-cfrg-gcmsiv-08](https://tools.ietf.org/html/draft-irtf-cfrg-gcmsiv-08))

## Add to your project

```xml
<dependency>
  <groupId>com.codahale</groupId>
  <artifactId>aes-gcm-siv</artifactId>
  <version>0.4.3</version>
</dependency>
```

*Note: module name for Java 9+ is `com.codahale.aesgcmsiv`.*

## Use the thing

```java
import com.codahale.aesgcmsiv.AEAD;
import java.nio.charset.StandardCharsets;
import java.util.Optional;

class Example {
  void roundTrip() {
    final AEAD aead = new AEAD(decodeHex("ee8e1ed9ff2540ae8f2ba9f50bc2f27c"));
    
    final byte[] plaintext = "Hello world".getBytes(StandardCharsets.UTF_8);
    final byte[] data = "example".getBytes(StandardCharsets.UTF_8);
   
    // automatically generates a nonce
    final byte[] ciphertext = aead.seal(plaintext, data);
    
    // automatically parses the nonce from the ciphertext
    final Optional<byte[]> result = aead.open(ciphertext, data);

    System.out.println(result.map(String::new));
  } 
}
```

`AEAD` also has versions of `seal` and `open` which support pre-generated nonces.

## License

Copyright © 2017 Coda Hale

Distributed under the Apache License 2.0.


CHANGELOG.md
# Change Log

## v0.4.3: 2018-09-26

* Remove JSR-305 annotations.
* Add automatic module name.

## v0.4.2: 2017-10-28

* Improved buffer reuse.
* Internal refactorings.

## v0.4.1: 2017-06-15

* Reduced allocation and copying for `POLYVAL` calculation.
* Improved performance very slightly, but now on par with AES-GCM.

## v0.4.0: 2017-06-14

* Moved to `byte[]`-based API.
* Removed okio dependency.

## v0.3.1: 2017-06-14

* Sequentialized reads and writes.
* Optimized keystream generation.
* Optimized counter operations.
* Improved performance by an additional ~10%.

## v0.3.0: 2017-06-12

* Added a 64-bit implementation of `POLYVAL`, doubling performance.
* Moved AES usage to JCE.
* Removed Bouncy Castle dependency.

## v0.2.2: 2017-06-11

* Eliminated unnecessary AES key scheduling.

## v0.2.1: 2017-05-26

* Added JSR 305 annotations.

## v0.2.0: 2017-05-25

* Added automatic nonce management.
* Small internal refactorings.

## v0.1.0: 2017-05-22

* Initial release.

src/test/java/com/codahale/aesgcmsiv/perf/PerfHarness.java
/*
 * Copyright © 2017 Coda Hale (coda.hale@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.codahale.aesgcmsiv.perf;

import com.codahale.aesgcmsiv.AEAD;

public class PerfHarness {

  @SuppressWarnings("ResultOfMethodCallIgnored")
  public static void main(String[] args) {
    final AEAD aead = new AEAD(new byte[16]);
    final byte[] nonce = new byte[12];
    final byte[] plaintext = new byte[1024];
    final byte[] data = new byte[0];

    for (int i = 0; i < 1_000_000; i++) {
      aead.seal(nonce, plaintext, data);
    }
  }
}


src/test/java/com/codahale/aesgcmsiv/benchmarks/Benchmarks.java
/*
 * Copyright © 2017 Coda Hale (coda.hale@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.codahale.aesgcmsiv.benchmarks;

import com.codahale.aesgcmsiv.AEAD;
import java.util.Optional;
import java.util.concurrent.TimeUnit;
import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;

@State(Scope.Benchmark)
@OutputTimeUnit(TimeUnit.MICROSECONDS)
@BenchmarkMode(Mode.AverageTime)
public class Benchmarks {

  private final byte[] key = new byte[16];
  private final byte[] nonce = new byte[12];
  private final byte[] plaintext = new byte[1024];
  private final byte[] data = new byte[0];
  private final AEAD aead = new AEAD(key);

  private byte[] gcmCiphertext = new byte[0];
  private byte[] sivCiphertext = new byte[0];

  @Setup
  public void setup() throws Exception {
    this.gcmCiphertext = aes_GCM_Encrypt();
    this.sivCiphertext = aes_GCM_SIV_Encrypt();
  }

  @Benchmark
  public byte[] aes_GCM_SIV_Encrypt() {
    return aead.seal(nonce, plaintext, data);
  }

  @Benchmark
  public Optional<byte[]> aes_GCM_SIV_Decrypt() {
    return aead.open(nonce, sivCiphertext, data);
  }

  @Benchmark
  public byte[] aes_GCM_Encrypt() throws Exception {
    final Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
    final GCMParameterSpec gcmSpec = new GCMParameterSpec(128, nonce);
    final SecretKeySpec keySpec = new SecretKeySpec(key, "AES");
    cipher.init(Cipher.ENCRYPT_MODE, keySpec, gcmSpec);
    return cipher.doFinal(plaintext);
  }

  @Benchmark
  public Optional<byte[]> aes_GCM_Decrypt() throws Exception {
    try {
      final Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
      final GCMParameterSpec gcmSpec = new GCMParameterSpec(128, nonce);
      final SecretKeySpec keySpec = new SecretKeySpec(key, "AES");
      cipher.init(Cipher.DECRYPT_MODE, keySpec, gcmSpec);
      return Optional.of(cipher.doFinal(gcmCiphertext));
    } catch (BadPaddingException e) {
      return Optional.empty();
    }
  }
}


src/main/java/com/codahale/aesgcmsiv/Bytes.java
/*
 * Copyright © 2017 Coda Hale (coda.hale@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.codahale.aesgcmsiv;

interface Bytes {

  static void putInt(int n, byte[] b) {
    b[0] = (byte) (n);
    b[1] = (byte) (n >> 8);
    b[2] = (byte) (n >> 16);
    b[3] = (byte) (n >> 24);
  }

  static void putLong(long n, byte[] b, int offset) {
    final int hi = (int) (n & 0xffffffffL);
    final int lo = (int) (n >> 32);
    b[offset] = (byte) hi;
    b[++offset] = (byte) (hi >> 8);
    b[++offset] = (byte) (hi >> 16);
    b[++offset] = (byte) (hi >> 24);
    b[++offset] = (byte) lo;
    b[++offset] = (byte) (lo >> 8);
    b[++offset] = (byte) (lo >> 16);
    b[++offset] = (byte) (lo >> 24);
  }

  static int getInt(byte[] b, int offset) {
    int n = b[offset] & 0xff;
    n |= (b[++offset] & 0xff) << 8;
    n |= (b[++offset] & 0xff) << 16;
    n |= b[++offset] << 24;
    return n;
  }

  static long getLong(byte[] b, int offset) {
    int lo = b[offset] & 0xff;
    lo |= (b[++offset] & 0xff) << 8;
    lo |= (b[++offset] & 0xff) << 16;
    lo |= b[++offset] << 24;

    int hi = b[++offset] & 0xff;
    hi |= (b[++offset] & 0xff) << 8;
    hi |= (b[++offset] & 0xff) << 16;
    hi |= b[++offset] << 24;

    return ((hi & 0xffffffffL) << 32) | (lo & 0xffffffffL);
  }
}


src/main/java/com/codahale/aesgcmsiv/Polyval.java
/*
 * Copyright © 2017 Coda Hale (coda.hale@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.codahale.aesgcmsiv;

// An implementation of POLYVAL based on GHASH because even BoringSSL doesn't have a POLYVAL impl.
// Does its own byte-order conversion to avoid confusion.
final class Polyval {

  private static final long E = 0xe100000000000000L;
  private static final int E1 = 0xe1000000;
  private final long h0;
  private final long h1;
  private long s0;
  private long s1;

  // mulX_GHASH, basically
  Polyval(byte[] h) {
    int v3 = Bytes.getInt(h, 0);
    int v2 = Bytes.getInt(h, 4);
    int v1 = Bytes.getInt(h, 8);
    int v0 = Bytes.getInt(h, 12);

    int b = v0;
    v0 = b >>> 1;
    int c = b << 31;
    b = v1;
    v1 = (b >>> 1) | c;
    c = b << 31;
    b = v2;
    v2 = (b >>> 1) | c;
    c = b << 31;
    b = v3;
    v3 = (b >>> 1) | c;
    v0 ^= (b << 31 >> 8 & E1);

    this.h0 = ((v0 & 0xffffffffL) << 32) | (v1 & 0xffffffffL);
    this.h1 = ((v2 & 0xffffffffL) << 32) | (v3 & 0xffffffffL);
  }

  void update(byte[] b) {
    final int extra = b.length % AEAD.AES_BLOCK_SIZE;
    for (int i = 0; i < b.length - extra; i += AEAD.AES_BLOCK_SIZE) {
      updateBlock(b, i);
    }

    if (extra != 0) {
      final byte[] block = new byte[AEAD.AES_BLOCK_SIZE];
      System.arraycopy(b, b.length - extra, block, 0, extra);
      updateBlock(block, 0);
    }
  }

  @SuppressWarnings("Duplicates")
  void updateBlock(byte[] b, int offset) {
    long v0 = h0;
    long v1 = h1;
    long z0 = 0;
    long z1 = 0;

    long x0 = s1 ^ Bytes.getLong(b, offset);
    long x1 = s0 ^ Bytes.getLong(b, offset + 8);

    // breaking this up into two duplicate loops is faster

    for (int i = 0; i < 64; i++) {
      long m = x1 >> 63;
      z0 ^= v0 & m;
      z1 ^= v1 & m;
      m = v1 << 63 >> 63;
      final long c = v0 & 1L;
      v0 >>>= 1;
      v1 = v1 >>> 1 | c << 63;
      v0 ^= E & m;
      x1 <<= 1;
    }

    for (int i = 64; i < 127; i++) {
      long m = x0 >> 63;
      z0 ^= v0 & m;
      z1 ^= v1 & m;
      m = v1 << 63 >> 63;
      final long c = v0 & 1L;
      v0 >>>= 1;
      v1 = v1 >>> 1 | c << 63;
      v0 ^= E & m;
      x0 <<= 1;
    }

    final long m = x0 >> 63;
    this.s0 = (z0 ^ (v0 & m));
    this.s1 = (z1 ^ (v1 & m));
  }

  void digest(byte[] d) {
    Bytes.putLong(s1, d, 0);
    Bytes.putLong(s0, d, 8);
  }
}


src/main/java/com/codahale/aesgcmsiv/AEAD.java
/*
 * Copyright © 2017 Coda Hale (coda.hale@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.codahale.aesgcmsiv;

import java.security.InvalidKeyException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.Arrays;
import java.util.Optional;
import javax.crypto.Cipher;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.ShortBufferException;
import javax.crypto.spec.SecretKeySpec;

/**
 * An AES-GCM-SIV AEAD instance.
 *
 * @see <a
 *     href="https://tools.ietf.org/html/draft-irtf-cfrg-gcmsiv-08">draft-irtf-cfrg-gcmsiv-08</a>
 * @see <a href="https://eprint.iacr.org/2017/168">AES-GCM-SIV: Specification and Analysis</a>
 */
public class AEAD {

  static final int AES_BLOCK_SIZE = 16;
  private static final int NONCE_SIZE = 12;

  private final Cipher aes;
  private final SecureRandom random;
  private final boolean aes128;

  /**
   * Creates a new {@link AEAD} instance with the given key.
   *
   * @param key the secret key; must be 16 or 32 bytes long
   */
  public AEAD(byte[] key) {
    if (key.length != 16 && key.length != 32) {
      throw new IllegalArgumentException("Key must be 16 or 32 bytes long");
    }
    this.aes = newAES(key);
    this.random = new SecureRandom();
    this.aes128 = key.length == 16;
  }

  /**
   * Encrypts the given plaintext.
   *
   * @param nonce a 12-byte random nonce
   * @param plaintext a plaintext message (may be empty)
   * @param data authenticated data (may be empty)
   * @return the encrypted message
   */
  public byte[] seal(byte[] nonce, byte[] plaintext, byte[] data) {
    if (nonce.length != NONCE_SIZE) {
      throw new IllegalArgumentException("Nonce must be 12 bytes long");
    }
    final byte[] authKey = subKey(0, 1, nonce);
    final Cipher encAES = newAES(subKey(2, aes128 ? 3 : 5, nonce));
    final byte[] tag = hash(encAES, authKey, nonce, plaintext, data);
    final byte[] output = new byte[plaintext.length + tag.length];
    aesCTR(encAES, tag, plaintext, output);
    System.arraycopy(tag, 0, output, plaintext.length, tag.length);
    return output;
  }

  /**
   * Encrypts the given plaintext, using a random nonce. Prepends the nonce to the resulting
   * ciphertext.
   *
   * @param plaintext a plaintext message (may be empty)
   * @param data authenticated data (may be empty)
   * @return the random nonce and the encrypted message
   */
  public byte[] seal(byte[] plaintext, byte[] data) {
    final byte[] nonce = new byte[NONCE_SIZE];
    random.nextBytes(nonce);

    final byte[] ciphertext = seal(nonce, plaintext, data);
    final byte[] output = new byte[nonce.length + ciphertext.length];
    System.arraycopy(nonce, 0, output, 0, nonce.length);
    System.arraycopy(ciphertext, 0, output, nonce.length, ciphertext.length);
    return output;
  }

  /**
   * Decrypts the given encrypted message.
   *
   * @param nonce the 12-byte random nonce used to encrypt the message
   * @param ciphertext the returned value from {@link #seal(byte[], byte[], byte[])}
   * @param data the authenticated data used to encrypt the message (may be empty)
   * @return the plaintext message
   */
  public Optional<byte[]> open(byte[] nonce, byte[] ciphertext, byte[] data) {
    if (nonce.length != NONCE_SIZE) {
      throw new IllegalArgumentException("Nonce must be 12 bytes long");
    }

    final byte[] c = new byte[ciphertext.length - AES_BLOCK_SIZE];
    final byte[] tag = new byte[AES_BLOCK_SIZE];
    System.arraycopy(ciphertext, 0, c, 0, c.length);
    System.arraycopy(ciphertext, c.length, tag, 0, tag.length);

    final byte[] authKey = subKey(0, 1, nonce);
    final Cipher encAES = newAES(subKey(2, aes128 ? 3 : 5, nonce));
    aesCTR(encAES, tag, c, c);
    final byte[] actual = hash(encAES, authKey, nonce, c, data);

    if (MessageDigest.isEqual(tag, actual)) {
      return Optional.of(c);
    }
    return Optional.empty();
  }

  /**
   * Decrypts the given encrypted message.
   *
   * @param ciphertext the returned value from {@link #seal(byte[], byte[])}
   * @param data the authenticated data used to encrypt the message (may be empty)
   * @return the plaintext message
   */
  public Optional<byte[]> open(byte[] ciphertext, byte[] data) {
    if (ciphertext.length < NONCE_SIZE) {
      return Optional.empty();
    }

    final byte[] nonce = new byte[NONCE_SIZE];
    final byte[] c = new byte[ciphertext.length - NONCE_SIZE];
    System.arraycopy(ciphertext, 0, nonce, 0, nonce.length);
    System.arraycopy(ciphertext, nonce.length, c, 0, c.length);

    return open(nonce, c, data);
  }

  private byte[] hash(Cipher aes, byte[] h, byte[] nonce, byte[] plaintext, byte[] data) {
    final Polyval polyval = new Polyval(h);
    polyval.update(data); // hash data with padding
    polyval.update(plaintext); // hash plaintext with padding

    // hash data and plaintext lengths in bits with padding
    final byte[] block = new byte[AES_BLOCK_SIZE];
    Bytes.putLong((long) data.length * 8, block, 0);
    Bytes.putLong((long) plaintext.length * 8, block, 8);
    polyval.updateBlock(block, 0);

    polyval.digest(block);
    for (int i = 0; i < nonce.length; i++) {
      block[i] ^= nonce[i];
    }
    block[block.length - 1] &= (byte) ~0x80;

    // encrypt polyval hash to produce tag
    try {
      aes.update(block, 0, block.length, block, 0);
    } catch (ShortBufferException e) {
      throw new IllegalStateException(e);
    }
    return block;
  }

  private byte[] subKey(int ctrStart, int ctrEnd, byte[] nonce) {
    final byte[] counter = new byte[AES_BLOCK_SIZE];
    System.arraycopy(nonce, 0, counter, counter.length - nonce.length, nonce.length);
    final byte[] key = new byte[(ctrEnd - ctrStart + 1) * 8];
    final byte[] block = new byte[AES_BLOCK_SIZE];
    for (int i = ctrStart; i <= ctrEnd; i++) {
      Bytes.putInt(i, counter);
      try {
        aes.update(counter, 0, AES_BLOCK_SIZE, block, 0);
      } catch (ShortBufferException e) {
        throw new IllegalStateException(e);
      }
      System.arraycopy(block, 0, key, (i - ctrStart) * 8, 8);
    }
    return key;
  }

  private void aesCTR(Cipher aes, byte[] tag, byte[] input, byte[] output) {
    final byte[] counter = Arrays.copyOf(tag, tag.length);
    counter[counter.length - 1] |= (byte) 0x80;
    final byte[] k = new byte[AES_BLOCK_SIZE];
    for (int i = 0; i < input.length; i += AES_BLOCK_SIZE) {
      // encrypt counter to produce keystream
      try {
        aes.update(counter, 0, counter.length, k, 0);
      } catch (ShortBufferException e) {
        throw new IllegalStateException(e);
      }

      // xor input with keystream
      final int len = Math.min(AES_BLOCK_SIZE, input.length - i);
      for (int j = 0; j < len; j++) {
        final int idx = i + j;
        output[idx] = (byte) (input[idx] ^ k[j]);
      }

      // increment counter
      int j = 0;
      while (j < 4 && ++counter[j] == 0) {
        j++;
      }
    }
  }

  private Cipher newAES(byte[] key) {
    try {
      final Cipher cipher = Cipher.getInstance("AES/ECB/NoPadding");
      cipher.init(Cipher.ENCRYPT_MODE, new SecretKeySpec(key, "AES"));
      return cipher;
    } catch (NoSuchAlgorithmException | NoSuchPaddingException | InvalidKeyException e) {
      throw new IllegalStateException(e);
    }
  }
}
